feat(dashboard): New analytics page UI fixes NV-6444

scopsy opened the pull request:
### What changed? Why was the change needed?

<!-- Also include any relevant links, such as Linear tickets, Slack discussions, or design documents. -->

### Screenshots

<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

<!-- A link to a dependent pull request  -->

### Special notes for your reviewer

<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>

scopsy committed:
Add chart primitives and Recharts integration

Introduces a new chart.tsx component with reusable chart primitives using Recharts for the dashboard app. Adds the 'recharts' dependency, updates index.css with chart color variables, and updates package and lock files to support charting features.

scopsy committed:
Update feature-flags.ts

scopsy committed:
Add new Home page and navigation link

Introduces a new Home page component, updates routing to include the Home page, and adds a conditional navigation link for the Home page based on a feature flag. This provides a dedicated landing experience for users and improves navigation structure.

scopsy committed:
Add randomized welcome messages to home page

Replaces the static greeting in the dashboard home page with a randomly selected message from a predefined list to enhance user experience. The subtitle is now stored in a constant for easier management.

scopsy committed:
Add AnalyticsCard component and integrate on home page

Introduces a reusable AnalyticsCard component with examples and documentation. Integrates the card into the dashboard home page to display key metrics such as messages delivered, active subscribers, interaction rate, and average messages per subscriber.

scopsy committed:
Update home-page.tsx

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Add delivery trends chart to home page

Introduces a new DeliveryTrendsChart component displaying daily message volume by channel using Recharts. Updates home page to include the chart and adds custom hover styles for bar elements in index.css.

scopsy committed:
Add new charts and update dashboard layout

Introduces InteractionTrendChart and WorkflowsByVolume components, updates DeliveryTrendsChart styling and data, and refactors chart primitives for improved flexibility. The home page layout now displays three charts in a grid. Removes unused Recharts bar hover styles from index.css.

scopsy committed:
fix arrows and icons

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
fix: tooltips

scopsy committed:
Refactor navigation and update chart/welcome styles

Replaces GettingStartedMenuItem with HomeMenuItem to support feature flag for new home page, updates references in side navigation, and adjusts logic for badge and icon display. Refactors ProgressSection to conditionally render WelcomeHeader based on new home page flag. Updates chart and tooltip components with new class names and color variables for improved consistency. Home page now includes ProgressSection and reorganizes resource lists.

scopsy committed:
Update dashboard UI with new shadow and layout styles

Replaces card and analytics component shadows with 'shadow-box-xs' and removes borders for a cleaner look. Refactors home page layout to use tighter grid gaps, adds a separator, and introduces a 'Things to do' section with improved onboarding progress header. Adds 'box-xs' shadow to Tailwind config for consistent styling.

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

djabarovgeorge committed:
feat(api-service): implement get charts (#8869)

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Update progress-section.tsx

scopsy committed:
Add workflow volume chart to dashboard and API

Introduces a new 'Top workflows by volume' chart to the dashboard, including backend support for fetching workflow run volumes, new DTOs, and API endpoints. Updates the dashboard UI to display the new chart, adds a React hook for fetching chart data, and enhances chart components with loading skeletons and improved tooltips.

scopsy committed:
Add messages delivered analytics to dashboard and API

Introduces a new 'messages delivered' report type to both the API and dashboard. Adds DTOs, use case, and repository method for fetching delivered message counts for current and previous periods. Updates dashboard components to display this metric in the analytics card, including loading states and trend calculation. Removes fallback chart data from delivery trends and workflows by volume components.

scopsy committed:
Add active subscribers chart to analytics dashboard

Introduces support for an 'Active Subscribers' chart in both the API and dashboard. Adds new DTOs, use case, and repository method to fetch and display active subscriber counts for current and previous periods. Updates the dashboard to request and render this new metric alongside existing analytics.

scopsy committed:
Add avg messages per subscriber chart to activity

Introduces support for the 'Average Messages per Subscriber' chart in the activity module. Adds new DTOs, use case, and repository method to calculate and return this metric for both current and previous periods. Updates API and dashboard types, as well as the home page to display the new chart data.

scopsy committed:
Add workflow runs metric chart to activity dashboard

Introduces a new workflow runs metric chart to the activity dashboard. Adds DTOs, use case, repository method, and integrates the metric into the API and dashboard UI. This enables tracking and comparison of workflow runs between current and previous periods.

scopsy committed:
Add interaction trend chart to activity analytics

Introduces the interaction trend chart to the activity analytics, including backend data aggregation, DTOs, and dashboard integration. Adds new API endpoints, data types, and UI components to display message sent, seen, read, and snoozed trends. Updates relevant enums, types, and chart rendering logic to support the new report type.

scopsy committed:
Add 'messageArchived' to interaction trend chart

Introduces support for tracking and displaying 'messageArchived' events in the interaction trend chart across API, dashboard, and DTOs. Updates data models, chart configuration, and backend aggregation to include the new metric.

scopsy committed:
Update trace-log.repository.ts

scopsy committed:
Update delivery-trends-chart.tsx

scopsy committed:
Update workflows-by-volume.tsx

scopsy committed:
Add date range filter to workflow runs stats on home page

Introduces a date range filter for workflow runs metrics, allowing users to select between 7, 30, or 90 days based on their subscription tier. Disabled options prompt an upgrade tooltip. The stats and charts now update according to the selected period, and the default range is determined by the user's plan and organization creation date.

scopsy committed:
Update get-charts.usecase.ts

scopsy committed:
Update activity.controller.ts

scopsy committed:
Update workflows-by-volume.tsx

scopsy committed:
Refactor and modularize home page charts and components

Moved and refactored chart components into a dedicated home-page/charts directory, introducing reusable chart wrappers, dummy data generators, and type definitions for chart data. Added new modular components for analytics, resources, stats, and welcome header, along with supporting constants and hooks for date filtering and metric data. Updated home page logic to use these new components and structure, improving maintainability and extensibility.

scopsy committed:
Add chat data to delivery trends chart tooltip

Extended the DeliveryTooltip component to include 'chat' as a delivery channel, displaying its value, color, and icon. Also removed redundant comments from chart-dummy-data.tsx and chart-types.ts for code clarity.

scopsy committed:
Update top-level-stats.tsx

scopsy committed:
Add total interactions chart to analytics dashboard

Introduces the 'total interactions' metric to the analytics dashboard, including backend support, DTOs, repository queries, and frontend integration. This enables tracking and displaying the total number of user interactions (seen, read, snoozed, archived) over a selected period, with comparison to the previous period.

scopsy committed:
Add workflow runs trend chart feature

Introduces the WorkflowRunsTrend chart, including DTOs, command, use case, and repository method to fetch and aggregate workflow run statuses over time. Updates the GetCharts use case and module registration to support the new report type.

scopsy committed:
analytics papge

scopsy committed:
Refactor analytics components and logic

Moved analytics-related charts, components, constants, and hooks from the home-page directory to a new analytics directory. Introduced new hooks for date filtering and metric data processing, and updated the analytics page to use these abstractions. The home page now uses a fixed 30-day period for metrics, while the analytics page supports selectable date ranges and improved chart/metric display.

scopsy committed:
Add workflow runs trend chart to analytics dashboard

Introduces a new WorkflowRunsTrendChart component to display workflow run trends, updates API types and chart dummy data, and integrates the chart into the analytics page. Also updates constants and chart types to support the new chart.

scopsy committed:
Add active subscribers trend and provider volume charts

Introduces backend and frontend support for two new analytics charts: Active Subscribers Trend and Top Providers by Volume. Adds DTOs, use cases, repository methods, and API wiring for both charts in the API, and implements corresponding React components and types in the dashboard. Updates chart configuration and analytics page to display the new charts.

scopsy committed:
Update workflow-diff.operation.ts

scopsy committed:
Update providers-by-volume.tsx

scopsy committed:
Update providers-by-volume.tsx

scopsy committed:
a

scopsy committed:
Update workflow-runs-trend-chart.tsx

scopsy committed:
Adjust chart x-axis ticks and date range handling

Updated x-axis tick frequency in ActiveSubscribersTrendChart to show every 4th tick. Modified DeliveryTrendsChart and InteractionTrendChart to use the second data point as the first date, and simplified x-axis ticks in InteractionTrendChart to only show the first and last dates. These changes improve chart readability and consistency.

scopsy committed:
Update layout and chart config on home page

Improved the ResourcesSection layout by adjusting flex properties for better responsiveness. Updated home page to memoize the 30-day date calculation and changed chart fetching to use only the WORKFLOW_RUNS_METRIC report type.

scopsy committed:
Refactor analytic log repositories to remove try/catch blocks

Removed redundant try/catch error handling from analytic log repository methods, allowing errors to propagate naturally. Also cleaned up unused imports and removed unnecessary query filters for null or empty subscriber and provider IDs. This simplifies the code and improves maintainability.

scopsy committed:
Rename analytics constants file and update imports

Renamed home-page.ts to analytics-page.consts.ts for clarity and updated all relevant imports. Minor formatting changes were made in home-page.tsx.

scopsy committed:
Update HomeMenuItem and home page stats rendering logic

Refines the condition for rendering HomeMenuItem to require both the new home page feature flag and user metadata. Also updates the home page to only show TopLevelStats if loading or there is data, preventing empty stats display.

scopsy committed:
Update analytics-card.tsx

scopsy committed:
Add animated number component to dashboard stats

Introduced a reusable AnimatedNumber component using @number-flow/react for animating numeric values in dashboard stats and analytics cards. Updated top-level stats and analytics card components to use the new animation, and added utility functions for compact number formatting and parsing. Minor UI spacing adjustments were made for improved layout.

scopsy committed:
Add useDelayedLoading hook to improve loading UX

Introduced a new useDelayedLoading hook to delay the display of loading skeletons, preventing unnecessary flashes for quick API responses. Updated TopLevelStats to use this hook for a smoother user experience.

scopsy committed:
Update home-page.tsx

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

djabarovgeorge commented on libs/application-generic/src/services/analytic-logs/workflow-run/workflow-run.repository.ts:
```diff
@@ -453,4 +453,263 @@ export class WorkflowRunRepository extends LogRepository<typeof workflowRunSchem
+        AND organization_id = {organizationId:String}
+        AND created_at >= {startDate:DateTime64(3)}
+        AND created_at <= {endDate:DateTime64(3)}
+        AND status = 'success'
```
Are we interested only in the "success" workflow runs? what if the user has a lot of pending workflows? (in-progress or delayed/digested)

scopsy replied:
Yeah, that's valid

scopsy committed:
Refactor analytics data fetching and loading states

Separated metrics and charts data fetching in the analytics page for clearer data flow and improved loading state management. Updated components to use a delayed loading hook for smoother skeleton transitions.

djabarovgeorge commented on libs/application-generic/src/services/analytic-logs/workflow-run/workflow-run.repository.ts:
```diff
@@ -453,4 +453,263 @@ export class WorkflowRunRepository extends LogRepository<typeof workflowRunSchem
   }
+
+  async getWorkflowVolumeData(
+    environmentId: string,
+    organizationId: string,
```
i wonder if we can add another layer of protection in order to make sure that those fields are passed to the clickhouseService.query request.
i cannot think of a good way to do so in compilation time, perhaps we should do it at least in runtime.

djabarovgeorge commented on libs/application-generic/src/services/analytic-logs/workflow-run/workflow-run.repository.ts:
```diff
@@ -453,4 +453,263 @@ export class WorkflowRunRepository extends LogRepository<typeof workflowRunSchem
+        'Failed to get workflow volume data'
+      );
+
+      throw error;
```
i am not sure we need to kill all of the analytic requests in case one of them failed.
or at least the user of the repository needs to decide on that result.

scopsy replied:
I aggree, forgot to remove it

scopsy committed:
Improve AnimatedNumber loading and update workflow query

AnimatedNumber now only resets to 0 on the initial load and avoids updating during subsequent loading states, improving UI behavior. In workflow-run.repository.ts, removed the 'status = success' filter from the workflow volume query and simplified error handling for cleaner code.

djabarovgeorge commented on libs/application-generic/src/services/analytic-logs/workflow-run/workflow-run.repository.ts:
```diff
@@ -453,4 +453,247 @@ export class WorkflowRunRepository extends LogRepository<typeof workflowRunSchem
+        AND created_at <= {endDate:DateTime64(3)}
+    `;
+
+    // Query for previous period
+    const previousPeriodQuery = `
+      SELECT count(DISTINCT external_subscriber_id) as count
+      FROM workflow_runs FINAL
```
non blocker: Instead of having a modular method, i think we could use the method twice instead.
once for the actual dates, then the second time for the previous dates.

scopsy committed:
Update changelog-cards.tsx

djabarovgeorge commented on apps/api/src/app/activity/usecases/build-active-subscribers-trend-chart/build-active-subscribers-trend-chart.command.ts:
```diff
@@ -0,0 +1,18 @@
+  @IsDefined()
+  endDate: Date;
+
+  static create(
+    data: Pick<BuildActiveSubscribersTrendChartCommand, 'environmentId' | 'organizationId' | 'startDate' | 'endDate'>
+  ) {
+    return Object.assign(new BuildActiveSubscribersTrendChartCommand(), data);
+  }
+}
```
🧹

scopsy committed:
Remove unused static create methods from chart commands

Deleted the static create methods from BuildActiveSubscribersTrendChartCommand and BuildProviderByVolumeChartCommand as they are no longer used.

djabarovgeorge commented on apps/dashboard/src/components/analytics/charts/chart-dummy-data.tsx:
will this be needed when we finish?

djabarovgeorge commented on apps/dashboard/src/components/analytics/charts/chart-wrapper.tsx:
```diff
@@ -0,0 +1,80 @@
+  error?: Error | null;
+  hasDataChecker: (data: T[]) => boolean;
+  loadingSkeleton: ReactNode;
+  dummyDataGenerator: () => T[];
```
will this will be needed when we finish?

scopsy replied:
It is, dummy data is used for the empty state mock image (the one that is faded out in the background if no data. is availabe)

djabarovgeorge commented on apps/dashboard/src/components/home-page/constants/home-page.ts:
can we share some of the code with the analytics page, or will it be removed later?

scopsy committed:
Remove unused CHART_CONFIG and related props

Deleted the CHART_CONFIG constant from home-page constants and removed its usage in the HomePage component. This cleans up unused configuration and simplifies the chart query setup.

scopsy committed:
Update side-navigation.tsx

scopsy committed:
Add analytics tooltips and info icons to dashboard charts

Introduces detailed tooltips for analytics metrics and charts by adding a new analytics-tooltips constant. Updates chart and card components to display an info icon with contextual tooltips. Refactors chart wrappers and analytics section to pass tooltip content, improving user understanding of analytics data. Also removes unrelated constants from analytics-page.consts.ts.

scopsy committed:
Refactor analytics tooltips to use HelpTooltipIndicator

Replaces inline tooltip implementations in chart and analytics card components with the reusable HelpTooltipIndicator component for consistency. Also updates tooltip content formatting and refines some tooltip text in analytics-tooltips constants.

scopsy committed:
Update analytics-section.tsx

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Improve chart empty state validation and tooltips

Refactored chart data validation logic into reusable utilities for date-based and volume-based charts. Enhanced empty state handling by adding customizable titles and tooltips, and updated analytics tooltips to clarify minimum data requirements for chart display.

scopsy committed:
Merge branch 'nv-6444-new-home-page-ui' of https://github.com/novuhq/novu into nv-6444-new-home-page-ui

scopsy committed:
Add AnalyticsUpgradeCtaIcon component and integrate

Introduces the AnalyticsUpgradeCtaIcon component for displaying an upgrade CTA with tooltip in the analytics section. Exports the new component and integrates it into the analytics page, passing it to the date filter hook for use in upgrade prompts.

scopsy committed:
Update get-charts.usecase.ts

scopsy committed:
Add instrumentation and 24h filter, remove Plain chat

Added the @InstrumentUsecase decorator to multiple chart usecases for improved observability. Introduced a 'Last 24 hours' option to the home page date filter. Removed the Plain live chat integration and related telemetry from the dashboard side navigation.

scopsy committed:
Add motion hover animation to inbox bell icons

Refactored InboxBell and InboxBellFilled components to use Framer Motion for hover animations, enhancing visual feedback. Updated workflow editor to use the custom InboxBell icon for IN_APP step type. Also renamed use-home-page-date-filter hook to use-analytics-page-date-filter for clarity and updated exports accordingly.

scopsy committed:
Refactor chart usecases and UI components

Removed unused organizationId and environmentId fields from BuildWorkflowByVolumeChartCommand. Updated percentage formatting in top-level stats to handle negative values. Fixed a typo in the changelog card container class. Cleaned up comments in chart usecases for previous period calculation.

scopsy committed:
Refactor metric data hook and update imports

Moved useMetricData hook to a shared location and updated all imports to reference the new path. Deleted redundant copies and removed analytics tooltips constants file. The formatNumber function now uses getCompactFormat for number formatting.

scopsy committed:
Update environment routing and feature flag handling

Replaces EnvironmentLevelCommand with EnvironmentCommand in the API. Adds CatchAllRoute as the index route in the dashboard router and updates CatchAllRoute to redirect to a new home page if the relevant feature flag is enabled. Refactors route matching logic for improved maintainability.

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Update chart config and fix MetricData import path

Added CHART_CONFIG with report types and timing constants to home-page constants. Also corrected the import path for MetricData in analytics-section.tsx to reference the correct hooks directory.

scopsy committed:
Remove new home page and related feature flag

Deleted the new home page, its components, constants, and hooks. Removed all references to the IS_NEW_HOME_PAGE_ENABLED feature flag and replaced it with IS_ANALYTICS_PAGE_ENABLED where appropriate. Updated navigation and routing logic to use the welcome page instead of the home page.

scopsy committed:
Merge branch 'next' into nv-6444-new-home-page-ui

scopsy committed:
Update pnpm-lock.yaml

scopsy committed:
Update package.json

scopsy committed:
Delete get-charts.e2e.ts
